Java/core/src/main/java/com/amazon/randomcutforest/tree/AbstractNodeStore.java

```diff
@@ -0,0 +1,527 @@
+        } else if (otherBox != null) {
+            return otherBox.probabilityOfCut(point);
+        } else {
+            BoundingBoxFloat box = getBox(node, pointStoreView);
+            return box.probabilityOfCut(point);
+        }
+    }
+
+    protected abstract void decreaseMassOfInternalNode(int node);
+
+    protected abstract void increaseMassOfInternalNode(int node);
+
+    protected void manageAncestorsAdd(Stack<int[]> path, float[] point, IPointStoreView<float[]> pointStoreview) {
+        while (!path.isEmpty()) {
+            int index = path.pop()[0];
+            increaseMassOfInternalNode(index);
+            if (nodeCacheFraction > 0.0) {
+                checkContainsAndRebuildBox(index, point, pointStoreview);
+                checkContainsAndAddPoint(index, point);
+            }
+        }
+    }
+
+    protected void manageAncestorsDelete(Stack<int[]> path, float[] point, IPointStoreView<float[]> pointStoreview) {
+        boolean resolved = false;
+        while (!path.isEmpty()) {
+            int index = path.pop()[0];
+            decreaseMassOfInternalNode(index);
+            if (nodeCacheFraction > 0.0 && !resolved) {
+                resolved = checkContainsAndRebuildBox(index, point, pointStoreview);
+            }
+        }
+    }
+
+    public Stack<int[]> getPath(int root, float[] point, boolean verbose) {
+        int node = root;
+        Stack<int[]> answer = new Stack<>();
+        answer.push(new int[] { root, 0 });
+        while (!isLeaf(node)) {
+            if (leftOf(node, point)) {
+                answer.push(new int[] { getLeftIndex(node), getRightIndex(node) });
+                node = getLeftIndex(node);
+            } else {
+                answer.push(new int[] { getRightIndex(node), getLeftIndex(node) });
+                node = getRightIndex(node);
+            }
+        }
+        return answer;
+    }
+
+    public abstract void deleteInternalNode(int index);
+
+    public int getLeafMass(int index) {
+        int y = (index - capacity);
+        Integer value = leafMass.get(y);
+        if (value != null) {
+            return value;
+        } else {
+            return 1;
+        }
+    }
+
+    public abstract int getMass(int index);
+
+    public int getPointIndex(int index) {
+        return index - capacity;
+    }
+
+    public boolean leftOf(int node, float[] point) {
+        return point[getCutDimension(node)] <= cutValue[node - 1];
+    }
+
+    public int getSibling(int node, int parent) {
+        int sibling = getLeftIndex(parent);
+        if (node == sibling) {
+            sibling = getRightIndex(parent);
+        }
+        return sibling;
+    }
+
+    public abstract void spliceEdge(int parent, int node, int newNode);
+
+    public abstract void replaceParentBySibling(int grandParent, int parent, int node);
+
+    public double dynamicScore2(int root, int ignoreMass, double[] point, IPointStoreView<float[]> pointStoreView,
```

Comment:
Do we need all of these scoring methods?

Reply:
No, the goal was to measure (sort of) had optimized computation vis-a-vis the traversals. It seems that the traversals can be optimized to be pretty much the same.

Reply:
Removed